parachain/primitives/ethereum/src/lib.rs

```diff
@@ -27,3 +39,182 @@ impl From<ethabi::Error> for DecodeError {
 		DecodeError::InvalidABI(err)
 	}
 }
+
+/// Complete block header id.
+#[derive(Clone, Copy, Default, Encode, Decode, PartialEq, RuntimeDebug)]
+pub struct HeaderId {
+	/// Header number.
+	pub number: u64,
+	/// Header hash.
+	pub hash: H256,
+}
+
+/// An Ethereum block header.
+#[derive(Clone, Default, Encode, Decode, PartialEq, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct Header {
+	/// Parent block hash.
+	pub parent_hash: H256,
+	/// Block timestamp.
+	pub timestamp: u64,
+	/// Block number.
+	pub number: u64,
+	/// Block author.
+	pub author: Address,
+
+	/// Transactions root.
+	pub transactions_root: H256,
+	/// Block ommers hash.
+	pub ommers_hash: H256,
+	/// Block extra data.
+	pub extra_data: Bytes,
+
+	/// State root.
+	pub state_root: H256,
+	/// Block receipts root.
+	pub receipts_root: H256,
+	/// Block bloom.
+	pub logs_bloom: Bloom,
+	/// Gas used for contracts execution.
+	pub gas_used: U256,
+	/// Block gas limit.
+	pub gas_limit: U256,
+
+	/// Block difficulty.
+	pub difficulty: U256,
+	/// Vector of post-RLP-encoded fields.
+	pub seal: Vec<Bytes>,
+}
+
+impl Header {
+	/// Compute hash of this header (keccak of the RLP with seal).
+	pub fn compute_hash(&self) -> H256 {
+		keccak_256(&self.rlp(true)).into()
+	}
+
+	/// Returns header RLP with or without seals.
+	fn rlp(&self, with_seal: bool) -> Bytes {
+		let mut s = RlpStream::new();
+		if with_seal {
+			s.begin_list(13 + self.seal.len());
+		} else {
+			s.begin_list(13);
+		}
+
+		s.append(&self.parent_hash);
+		s.append(&self.ommers_hash);
+		s.append(&self.author);
+		s.append(&self.state_root);
+		s.append(&self.transactions_root);
+		s.append(&self.receipts_root);
+		s.append(&EthBloom::from(self.logs_bloom.0));
+		s.append(&self.difficulty);
+		s.append(&self.number);
+		s.append(&self.gas_limit);
+		s.append(&self.gas_used);
+		s.append(&self.timestamp);
+		s.append(&self.extra_data);
+
+		if with_seal {
+			for b in &self.seal {
+				s.append_raw(b, 1);
+			}
+		}
+
+		s.out()
+	}
+}
+
+#[cfg(feature = "std")]
+big_array! { BigArray; }
+
+/// Logs bloom.
+#[derive(Clone, Debug, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct Bloom(#[cfg_attr(feature = "std", serde(with = "BigArray"))] [u8; 256]);
+
+impl<'a> From<&'a [u8; 256]> for Bloom {
+	fn from(buffer: &'a [u8; 256]) -> Bloom {
+		Bloom(*buffer)
+	}
+}
+
+impl PartialEq<Bloom> for Bloom {
+	fn eq(&self, other: &Bloom) -> bool {
+		self.0.iter().zip(other.0.iter()).all(|(l, r)| l == r)
+	}
+}
+
+impl Default for Bloom {
+	fn default() -> Self {
+		Bloom([0; 256])
+	}
+}
+
+
+#[cfg(test)]
+mod tests {
+
+	use super::*;
+	use hex_literal::hex;
+
+	#[test]
+	fn header_compute_hash_poa() {
```

Comment:
do we need this?

Reply:
Unclear... the test is intended to demonstrate that the `Header` definition is consensus-engine-agnostic. The `seal` field captures either `mixHash` + `nonce` for PoW, `step` + `signature` for PoA, etc. I kept the `seal` field because of this description: https://ethereum.stackexchange.com/questions/6093/what-does-it-mean-to-seal-a-block. Happy to change if you think it's better to make this PoW-specific

Reply:
ah right, nah that's fine then. is there anything in this PR that you think we should try get changed back in parity-bridges-common?

Reply:
No. I made no significant changes except adding tests